nexus/pkg/artifacts/downloader.go

```diff
@@ -0,0 +1,309 @@
+package artifacts
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/wandb/wandb/nexus/internal/filetransfer"
+	"github.com/wandb/wandb/nexus/internal/gql"
+	"github.com/wandb/wandb/nexus/pkg/env"
+)
+
+const BATCH_SIZE int = 10000
+const MAX_BACKLOG int = 10000
+
+type ArtifactDownloader struct {
+	// Resources
+	Ctx             context.Context
+	GraphqlClient   graphql.Client
+	DownloadManager *filetransfer.FileTransferManager
+	// Input
+	QualifiedName          string
+	DownloadRoot           *string
+	Recursive              *bool
+	AllowMissingReferences *bool
+	// Properties
+	UpstreamArtifacts   []gql.ArtifactByIDArtifact
+	DefaultDownloadRoot *string
+}
+
+func NewArtifactDownloader(
+	ctx context.Context,
+	graphQLClient graphql.Client,
+	downloadManager *filetransfer.FileTransferManager,
+	qualifiedName string,
+	downloadRoot *string,
+	recursive *bool,
+	allowMissingReferences *bool,
+) *ArtifactDownloader {
+	return &ArtifactDownloader{
+		Ctx:                    ctx,
+		GraphqlClient:          graphQLClient,
+		DownloadManager:        downloadManager,
+		QualifiedName:          qualifiedName,
+		DownloadRoot:           downloadRoot,
+		Recursive:              recursive,
+		AllowMissingReferences: allowMissingReferences,
+		UpstreamArtifacts:      []gql.ArtifactByIDArtifact{},
+		DefaultDownloadRoot:    nil,
+	}
+}
+
+func (ad *ArtifactDownloader) getArtifact() (attrs gql.ArtifactByNameProjectArtifact, rerr error) {
+	entityName, projectName, artifactName, err := parseArtifactQualifiedName(ad.QualifiedName)
+	if err != nil {
+		return gql.ArtifactByNameProjectArtifact{}, err
+	}
+	response, err := gql.ArtifactByName(
+		ad.Ctx,
+		ad.GraphqlClient,
+		entityName,
+		projectName,
+		artifactName,
+	)
+	if err != nil {
+		return gql.ArtifactByNameProjectArtifact{}, err
+	}
+	return *response.GetProject().GetArtifact(), nil
```

Comment:
This will cause a seg fault if the artifact doesn't exist or your don't have permission to access it. Check for `nil`.